contrib/dynamic_embedding/src/tde/ps.cpp

```diff
@@ -136,22 +142,30 @@ void PS::Evict(torch::Tensor ids_to_evict) {
         col_ids,
         os_ids_,
         tcb::span{
-            reinterpret_cast<uint8_t*>(data.data_ptr<float>()),
+            reinterpret_cast<uint8_t*>(data_shared_ptr->data_ptr<float>()),
             data_size * sizeof(float)},
         tcb::span{offsets.data(), offsets_size},
-        [&notification] { notification.Done(); });
+        [&notification, data_shared_ptr] { notification.Done(); });
   }
   notification.Wait();
 }
 
 void PS::SyncFetch(int64_t time) {
-  while (!fetch_notifications_.empty()) {
-    auto& [t, notification] = fetch_notifications_.front();
-    if (t != time && time >= 0) {
+  std::unique_lock<std::mutex> lock(
+      fetch_notifications_mutex_, std::defer_lock);
```

Comment:
normally , SyncFetch and Fetch would be called concurrently, because Fetch returns an asynchronous handler and then continues to the next fetch. The asynchronous handler will be waited on by other threads calling SyncFetch. Therefore, using the mu_ lock would cause SyncFetch and Fetch to be completely mutually exclusive, which means that Fetch cannot be executed while waiting for SyncFetch, leading to a decline in performance. Therefore, a finer-grained lock is created here to protect only the access to the notification.

Reply:
This makes sense, I can see why we want the granularity

Reply:
Oops, approved sorry